### PR TITLE
Implement Batch Verifier For Pubsub Objects

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

In gossip, the beacon node receives multiple objects over the wire and has to verify them separately. This brings about a 
significant load to the beacon node as the validator set in mainnet grows. Beacon nodes running on single/dual cores will eventually be insufficient due to to increased number of pairings that need to be performed. By using batch verification with randomized factors, we can reduce the number of miller loops performed with the result of reducing the CPU time required to verify `n` signatures. In the event batch verification fails due to an errant signature somewhere this method will fallback to verifying all signatures one by one. 

- [x] Implement verifier routine to regularly perform batch verifications of incoming signature sets.
- [x] Modify relevant gossip validators to allow them to provider signature sets for batch verification.
- [x] Add feature flag to gate this feature.
- [x] Add in relevant tests for this feature. 

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

File: beacon-chain/sync/batch_verifier.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+			log.Debugf("Verification of %s failed", message)
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationReject
+		}
+	}
+	return pubsub.ValidationAccept
+}
+
+func verifyBatch(verifierBatch []*signatureVerifier) []*signatureVerifier {
+	aggSet := verifierBatch[0].set
```
Comment (terencechain) [inline beacon-chain/sync/batch_verifier.go:80]: please check for empty `verifierBatch`

File: beacon-chain/sync/batch_verifier.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		aggSet = aggSet.Join(verifierBatch[i].set)
+	}
+	verified, err := aggSet.Verify()
+	switch {
+	case err != nil:
+		verificationErr = err
+	case !verified:
+		verificationErr = errors.New("batch signature verification failed")
+	}
+	for i := 0; i < len(verifierBatch); i++ {
```
Comment (terencechain) [inline beacon-chain/sync/batch_verifier.go:93]: This is a rare event, we can avoid sending nil over the channel in most cases by predefining an `hasError` variable or something

File: beacon-chain/sync/batch_verifier.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		verifierBatch[i].resChan <- verificationErr
+	}
+	verifierBatch = []*signatureVerifier{}
+	return verifierBatch
```
Comment (terencechain) [inline beacon-chain/sync/batch_verifier.go:97]: Feels anti pattern to initialize and return a fresh list. should we let the callers do this?

File: beacon-chain/sync/batch_verifier.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/sync/batch_verifier.go:93]: In that case we would lose the ability to inform the caller on the error no ?

**Review by terencechain — APPROVED** [review]

**Pull request merged**